File: listnet_python/src-octave/listwise_cost.m
function J = listwise_cost(y,z, list_id)
    ly = @(i) y(find(list_id == list_id(i)),:);
    lz = @(i) z(find(list_id == list_id(i)),:);
    
    j = @(i) (-sum(topp(ly(i)) .* log( topp(lz(i)) )));
 


    J = arrayfun(j,1:size(z,1));
    
    fprintf("listwise_cost(1) = %f\n", j(1));
    fprintf("listwise_cost(2) = %f\n", j(2));
    fprintf("listwise_cost(10) = %f\n", j(10));
    fprintf("listwise_cost(16) = %f\n", j(16));
    fprintf("listwise_cost(17) = %f\n", j(17));
    fprintf("len(listwise_cost) = %d\n", size(J,2));
end

File: src/learning/Listwise_cost.py
import numpy as np
from learning import Globals
from learning import topp
from learning import find
from learning import exposure


def listwise_cost(GAMMA, training_judgments, predictions, query_ids, prot_idx):
    """
    computes the loss in list-wise learning to rank
    it incorporates L which is the error between the training judgments and those
    predicted by a model and U which is the disparate exposure metric
    implementation of equation 6 in DELTR paper

    :param GAMMA: a float parameter tuning the disparate exposure metric
    :param training_judgments: containing the training judgments/ scores
    :param predictions: containing the predicted scores
    :param query_ids: list of query IDs
    :param prot_idx: list stating which item is protected or non-protected
    :return: a float value --> loss
    """
    data_per_query = lambda which_query, data: \
                                   find.find_items_per_group_per_query(data, query_ids, which_query, prot_idx)

    loss = lambda which_query:-np.dot(np.transpose(topp.topp(data_per_query(which_query, training_judgments)[0])),
                                          np.log(topp.topp(data_per_query(which_query, predictions)[0])))  # eq 2 from DELTR paper

    if Globals.ONLY_L:
        cost = lambda which_query: loss(which_query)

    if Globals.L_AND_U:
        cost = lambda which_query: GAMMA * exposure.exposure_diff(predictions, query_ids, which_query, prot_idx) ** 2 + loss(which_query)

    results = [cost(query) for query in query_ids]

    return np.asarray(results)
